modules/dmrpp_module/DmrppArray.cc

```diff
@@ -1336,6 +1337,203 @@ void DmrppArray::read_chunks_dio_unconstrained()
 }
 
 
+// Retrieve data from the linked blocks.  We don't need to use the super chunk technique
+// since the adjacent blocks are already combined. We just need to read the data
+// from each chunk, combine them and decompress the buffer if necessary.
+void DmrppArray::read_linked_blocks(){
+    unsigned int num_linked_blocks = this->get_total_linked_blocks();
+    if (num_linked_blocks <2)
+        throw BESInternalError("The number of linked blocks must be >1 to read the data.", __FILE__, __LINE__);
+
+    //Change to the total storage buffer size to just the compressed buffer size.
+    reserve_value_capacity_ll_byte(get_var_chunks_storage_size());
+ //cout <<"total_linked_block is: "<<num_linked_blocks <<endl;
+
+    vector<unsigned long long> accumulated_lengths;
+    accumulated_lengths.resize(num_linked_blocks);
+    vector<unsigned long long> individual_lengths;
+    individual_lengths.resize(num_linked_blocks);
+
+    // Here we cannot assume that the index of the linked block always increases
+    // in the loop of chunks so we use the linked block index.
+    // For the HDF4 case, we observe the index of the linked block is always consistent with the
+    //  chunk it loops, though.
+    for(const auto& chunk: get_immutable_chunks()) {
+        individual_lengths[chunk->get_linked_block_index()] = chunk->get_size();
+    }
+    accumulated_lengths[0] = 0;
+    for (unsigned int i = 1; i < num_linked_blocks; i++)
+        accumulated_lengths[i] = individual_lengths[i-1] + accumulated_lengths[i-1];
+
+    char *target_buffer = get_buf();
+
+    for(const auto& chunk: get_immutable_chunks()) {
+        chunk->read_chunk();
+        // cout<<"linked_block_index is: " << chunk->get_linked_block_index() <<endl;
+        // cout<<"accumlated_length: "<< accumulated_lengths[chunk->get_linked_block_index()] <<endl;
+        const char *source_buffer = chunk->get_rbuf();
+        memcpy(target_buffer + accumulated_lengths[chunk->get_linked_block_index()], source_buffer,chunk->get_size());
+    }
+
+    string filters_string = this->get_filters();
+    if (filters_string.find("deflate")!=string::npos) {
+        char *in_buf = get_buf();
+
+        char **destp = nullptr;
+        char *dest_deflate = nullptr;
+        unsigned long long out_buf_size = 0;
+        unsigned long long dest_len = get_var_chunks_storage_size();
+        unsigned long long src_len = get_var_chunks_storage_size();
+        dest_deflate = new char[dest_len];
```

Comment:
You could use a vector<char> here and then would not need the delete[] and would not need to guard against exceptions. vector has a data() method that returns a pointer to the vector.